src/Scrutor/ServiceCollectionExtensions.Decoration.cs

```diff
@@ -21,7 +21,35 @@ public static partial class ServiceCollectionExtensions
         {
             Preconditions.NotNull(services, nameof(services));
 
-            return services.DecorateDescriptors(typeof(TService), x => x.Decorate(typeof(TDecorator)));
+            if (typeof(TDecorator).IsInterface)
+            {
+                return DecorateUsingInterface<TService, TDecorator>(services);
+            }
+            else
+            {
+                return services.DecorateDescriptors(typeof(TService), x => x.Decorate(typeof(TDecorator)));
+            }
+        }
+
+        private static IServiceCollection DecorateUsingInterface<TService, TDecorator>(IServiceCollection services) where TDecorator : TService
+        {
+            if (typeof(TDecorator).IsGenericType)
+            {
+                var decoratorDescriptor = services.Where(service => HasSameTypeDefinition(service.ServiceType, typeof(TDecorator))).FirstOrDefault();
+                if (decoratorDescriptor == null)
+                    throw new MissingTypeRegistrationException(typeof(TDecorator).IsGenericType ? typeof(TDecorator).GetGenericTypeDefinition() : typeof(TDecorator));
+
+                return services.DecorateDescriptors(typeof(TService), x => x.Decorate(decoratorDescriptor.ImplementationType.MakeGenericType(typeof(TDecorator).GetGenericArguments().First())));
+            }
+            else
+            {
+                var decoratorDescriptor = services.Where(service => service.ServiceType == typeof(TDecorator)).FirstOrDefault();
+                if (decoratorDescriptor == null)
+                    throw new MissingTypeRegistrationException(typeof(TDecorator).IsGenericType ? typeof(TDecorator).GetGenericTypeDefinition() : typeof(TDecorator));
```

Comment:
...and always `false` here? 🤔

Reply:
as per : https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.firstordefault?view=net-5.0
"The default value for reference and nullable types is null."